File: backend/src/aerospike_cluster_manager_api/models/record.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

BinValue = Any


class GeoJSON(BaseModel):
    type: str  # "Point" | "Polygon" | "AeroCircle"
    coordinates: list[Any]


class RecordKey(BaseModel):
    namespace: str = Field(min_length=1, max_length=31)
    set: str = Field(default="", max_length=63)
    pk: str = Field(default="", max_length=1024)
    digest: str | None = None


class RecordMeta(BaseModel):
    generation: int = Field(ge=0)
    ttl: int = Field(ge=0)
    lastUpdateMs: int | None = None


class AerospikeRecord(BaseModel):
    key: RecordKey
    meta: RecordMeta
    bins: dict[str, BinValue]


class RecordListResponse(BaseModel):
    records: list[AerospikeRecord]
    total: int
    page: int
    pageSize: int
    hasMore: bool
    totalEstimated: bool = False


class RecordWriteRequest(BaseModel):
    key: RecordKey
    bins: dict[str, BinValue]
    ttl: int | None = None

File: ui/src/lib/types/record.ts
/**
 * Record-related types mirrored from backend Pydantic models.
 * See: backend/src/aerospike_cluster_manager_api/models/record.py
 */

export type BinValue = unknown

export type PkType = "auto" | "string" | "int" | "bytes"

export interface GeoJSON {
  type: string
  coordinates: unknown[]
}

export interface RecordKey {
  namespace: string
  set?: string
  pk?: string
  digest?: string | null
}

export interface RecordMeta {
  generation: number
  ttl: number
  lastUpdateMs?: number | null
}

export interface AerospikeRecord {
  key: RecordKey
  meta: RecordMeta
  bins: Record<string, BinValue>
  /** Operator-authored memo from cluster-manager metaDB. Null when no note is attached. */
  note?: string | null
}

export interface RecordListResponse {
  records: AerospikeRecord[]
  /**
   * Object count for the set, or `null` when it could not be determined
   * (ADR-0026). `null` is an explicit "unknown" — it must NOT be rendered as
   * "no records", and it must not disable pagination. Show the loaded count
   * instead ("~N+ records").
   */
  total: number | null
  page: number
  pageSize: number
  hasMore: boolean
  totalEstimated: boolean
}

export interface RecordWriteRequest {
  key: RecordKey
  bins: Record<string, BinValue>
  ttl?: number | null
  pkType?: PkType
}
